### PR TITLE
Handle clearing problematic user preferences

### PR DESCRIPTION
**Pull request by beckyvb**
Clear the url from a user's preferences if it's one that redirects them to a 404, 401, etc.
Provide an endpoint to manually clear a user's navigation preferences.

### CODE REVIEW

File: applications/dashboard/controllers/class.homecontroller.php [hunk 1]
```diff
@@ -97,6 +99,15 @@ public function fileNotFound() {
     }
 
     /**
+     * Clears the url from the user's navigation preferences.
+     */
+    private function clearNavigationPreferences() {
+        $userModel = new UserModel();
+        $uri = Gdn::request()->getRequestArguments('server')['REQUEST_URI'];
```
Comment (linc) [inline applications/dashboard/controllers/class.homecontroller.php]: `Gdn::request()->path()` ?

File: applications/dashboard/controllers/class.dashboardcontroller.php [hunk 2]
```diff
@@ -59,6 +59,23 @@ public function initialize() {
     }
 
     /**
+     * An endpoint to clear all the navigation preferences for a user.
+     *
+     * @param string $userID
+     */
+    public function clearNavigationPreferences($userID = '') {
+        if ($userID && $userID !== Gdn::session()->UserID) {
```
Comment (linc) [inline applications/dashboard/controllers/class.dashboardcontroller.php]: Wanna check for a `POST` here so we can't CSRF folks.

File: applications/dashboard/controllers/class.dashboardcontroller.php [hunk 3]
```diff
@@ -59,6 +59,23 @@ public function initialize() {
...
+            $this->permission('Garden.Users.Edit');
+        }
+
+        $userModel = new UserModel();
+        $userModel->clearNavigationPreferences($userID);
+
+        $this->informMessage(t('Navigation preferences cleared.'));
+        redirect('/');
```
Comment (linc) [inline applications/dashboard/controllers/class.dashboardcontroller.php]: I'd make this redirect conditional on `DELIVERY_TYPE_ALL` so we don't bounce a menu / ajax request and miss the inform message, which never actually renders as-is. Render blank if you want that to show up.

**Comment by DaazKu** [comment]
> My overriding comment is: Do we really need all the detection of clearSectionNavigationPreference() or can we just blow out the preference completely on any error? Would anyone even notice that lesser granularity?

I would. As soon as we will get used to the dashboard remembering where we were, having it not land you where it was supposed to just because you hit a 404 somewhere will be considered a "bug"

**Comment by beckyvb** [comment]
I agree with @DaazKu. I was going to wipe out all the preferences, but I thought it would be pretty bad user experience.

File: applications/dashboard/controllers/class.homecontroller.php (hunk 1 above)
Comment (beckyvb) [inline applications/dashboard/controllers/class.homecontroller.php]: That's `dashboard/home/filenotfound` here

Comment (DaazKu) [inline applications/dashboard/controllers/class.homecontroller.php]: Gdn::request()->requestURI();

Comment (beckyvb) [inline applications/dashboard/controllers/class.homecontroller.php]: Same, that's `dashboard/home/filenotfound` here (or whichever endpoint we're redirected to)

Comment (DaazKu) [inline applications/dashboard/controllers/class.homecontroller.php]: I know it is the same but I think it is the intended way of using the Request object.
Otherwhise why not just `$_SERVER['REQUEST_URI'];`. Way simpler.

**Comment by linc** [comment]
I will acquiesce to your additional code complexity at the toll of at least 2 inline comments.

**Pull request merged**